core/tee/tee_svc_cryp.c

```diff
@@ -3873,15 +3873,21 @@ TEE_Result syscall_asymm_verify(unsigned long state,
 		res = tee_alg_get_digest_size(hash_algo, &hash_size);
 		if (res != TEE_SUCCESS)
 			break;
-		/*
-		 * Depending on the DSA algorithm (NIST), the digital signature
-		 * output size may be truncated to the size of a key pair
-		 * (Q prime size). Q prime size must be less or equal than the
-		 * hash output length of the hash algorithm involved.
-		 */
-		if (data_len > hash_size) {
-			res = TEE_ERROR_BAD_PARAMETERS;
-			break;
+
+		if (data_len != hash_size) {
+			struct dsa_public_key *key = o->attr;
+
+			/*
+			 * Depending on the DSA algorithm (NIST), the
+			 * digital signature output size may be truncated
+			 * to the size of a key pair (Q prime size). Q
+			 * prime size must be less or equal than the hash
+			 * output length of the hash algorithm involved.
+			 */
+			if (data_len != crypto_bignum_num_bytes(key->q)) {
+				res = TEE_ERROR_BAD_PARAMETERS;
+				break;
+			}
```

Comment:
I don't really understand the constraints and where they come from (which spec?). More importantly: why are these checks needed in the first place? Shouldn't they be handled by `crypto_acipher_dsa_verify()`?

Reply:
The updated GP tests checks something with regards to this. I'll dig out the details.

Reply:
Shouldn't it be the below?:
```
		struct dsa_public_key *key = o->attr;

		if ((hash_size < data_len) ||
		    (hash_size < crypto_bignum_num_bytes(key->q)) {
				res = TEE_ERROR_BAD_PARAMETERS;
				break;
		}

Reply:
> I don't really understand the constraints and where they come from (which spec?). More importantly: why are these checks needed in the first place? Shouldn't they be handled by `crypto_acipher_dsa_verify()`?

In case `data_len` is bad `TEE_AsymmetricVerifyDigest()` should panic. If we skip this test and LTC is unhappy it will return `TEE_ERROR_SIGNATURE_INVALID` instead. You can also see FIPS 186-4 section 4.7.

The GP spec says that we SHOULD panic if `data_len != hash_size`, but that would break a few of the DSA tests in xtest where the `hash_size` is larger than possible `data_len`. So the compromise is in case `data_len != hash_size` check that it's not smaller than what makes sense. I discovered this while updating the GP test suite which has a few tests with too small `data_len`.

> Shouldn't it be the below?:...

I don't understand, what are you trying to catch with that?

Reply:
> > I don't really understand the constraints and where they come from (which spec?). More importantly: why are these checks needed in the first place? Shouldn't they be handled by `crypto_acipher_dsa_verify()`?
> 
> In case `data_len` is bad `TEE_AsymmetricVerifyDigest()` should panic. If we skip this test and LTC is unhappy it will return `TEE_ERROR_SIGNATURE_INVALID` instead. You can also see FIPS 186-4 section 4.7.

OK, makes sense, could you maybe add a comment? Like "`crypto_acipher_dsa_verify()` is expected to follow the FIPS spec and return "invalid" when data_len is "bad", but the GP spec mandates a different behavior (TA panic)"?

> The GP spec says that we SHOULD panic if `data_len != hash_size`, but that would break a few of the DSA tests in xtest where the `hash_size` is larger than possible `data_len`. So the compromise is in case `data_len != hash_size` check that it's not smaller than what makes sense. I discovered this while updating the GP test suite which has a few tests with too small `data_len`.

That's useful information too, worth adding to the comment IMO :)

Why not `data_len < crypto_bignum_num_bytes(key->q)` then? (`<` instead of `!=`)

Reply:
Fair enough, I'll add comments.

I don't see how a `data_len` between `hash_size` and `crypto_bignum_num_bytes(key->q)` makes sense.
If `data_len` isn't `hash_size` it has been truncated due to the size of `Q`, but then it should have a length consistent with the size of `Q` and nothing else.